### PR TITLE
Feature: Parquet Schema from JSON Schema

### PR DESCRIPTION
**Pull request by wilwade**
Problem
=======

We want it to be easier to build Parquet files (and thus the schema for those files) from existing schemas. In this case JSON Schemas

Solution
========

Change summary:
---------------
* Added field creation helper functions
* Added static function `ParquetSchema.fromJsonSchema` that takes in a Json Schema

Steps to Verify:
----------------
1. Examples are in the tests, but take a JSON schema and call `ParquetSchema.fromJsonSchema` with it and see what it does

### CODE REVIEW

File: lib/fields.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+export function createStringField(optional = true, fieldOptions: FieldDefinition = {}): FieldDefinition {
+    return { ...fieldOptions, optional, type: 'UTF8' };
+}
+
+export function createBooleanField(optional = true, fieldOptions: FieldDefinition = {}): FieldDefinition {
+    return { ...fieldOptions, optional, type: 'BOOLEAN' };
+}
+
+export function createIntField(size: 32 | 64, optional = true, fieldOptions: FieldDefinition = {}): FieldDefinition {
```
Comment (noxify) [inline lib/fields.ts:13]: What about `INT96`? 
 --> https://parquet.apache.org/docs/file-format/types/

**Comment by noxify** [comment]
Haven't checked it in detail, but added a comment to the `createIntField` function.
I will check it in detail on monday when I'm back at work :)

File: lib/fields.ts (hunk 1 above)
Comment (wilwade) [reply lib/fields.ts:13]: Good call out. I checked and according to the Parquet Thrift INT96 is "only used for legacy timestamps".

File: README.md [hunk 2]
```diff
@@ -98,6 +100,34 @@ var schema = new parquet.ParquetSchema({
 });
 ```
 
+### JSON Schema
```
Comment (noxify) [inline README.md:103]: Does it makes sense to provide also an example with the `create<Type>Field` functions?

With this PR we have - please correct me if I'm wrong - three options to create a schema:

* native w/o helper functions ( plain? )
* native w/ helper functions
* JSON Schema

```
var schema = new parquet.ParquetSchema({
  name: parquet.ParquetSchema.createStringField(),
  quantity: parquet.ParquetSchema.createIntField(64),
  price: parquet.ParquetSchema.createDoubleField(),
  date: parquet.ParquetSchema.createTimestampField(),
  in_stock: parquet.ParquetSchema.createBooleanField()
});
```

Comment (wilwade) [reply README.md:115]: Good idea. Added!

File: lib/fields.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+export function createStructField(fields: SchemaDefinition, optional = true): FieldDefinition {
+    return {
+        optional,
+        fields,
+    }
+}
+
+export function createStructListField(fields: SchemaDefinition, optional = true): FieldDefinition {
+    return {
```
Comment (noxify) [inline lib/fields.ts:41]: I found some time to test the changes locally.

It seems to work and the parquet file is generated successfully, but if I try to open the generated parquet file in an external parquet viewer ( in my case via this [PyCharm Plugin](https://github.com/benwatson528/intellij-avro-parquet-plugin) ) I got the following error message:

```
Invalid list type repeated group struct_field (LIST) {
  required group list {
    optional binary sub3 (STRING);
    optional binary sub4 (STRING);
    optional group sub5 {
      optional binary sub6 (STRING);
      optional binary sub7 (STRING);
    }
    optional group sub8 (LIST) {
      repeated group list {
        required binary element (STRING);
      }
    }
  }
}


```

If I change the return value to:

```js
  return {
    type: "LIST",
    optional,
    fields: {
      list: {
        repeated: true,
        fields,
      },
    },
  }
```

The error doesn't occur.

File: lib/fields.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        fields: {
+            list: {
+                fields,
+            },
+        },
+    }
+}
+
+export function createListField(type: ParquetType, optionalElement = false, optionalListField = true, elementOptions: FieldDefinition = {}): FieldDefinition {
+    return {
```
Comment (noxify) [inline lib/fields.ts:54]: same test case as for `createStructListField`.

The error message is the following:
```
Invalid list type repeated group array_field (LIST) {
  required group list {
    required binary element (STRING);
  }
}

```

If I change the return value to:

```js
  return {
    type: "LIST",

    optional: optionalListField,
    fields: {
      list: {
        repeated: true,
        fields: {
          element: {
            ...elementOptions,
            type,
            optional: optionalElement,
          },
        },
      },
    },
  }
```

The error doesn't occur.

**Review by noxify — COMMENTED** [review]
Steps to reproduce:

1. Create a new file `json-test.js` in the project root ( code below )
2. Run `npm run build`
3. Run `node json-test.js`
4. Drag&Drop the generated parquet file ( `<projectroot>/example.parquet` ) into the PyCharm Parquet Viewer plugin

```js
// json-test.js
const parquetjs = require('./dist/parquet')

const examplePath = './example.parquet'

const parquetSchema = parquetjs.ParquetSchema.fromJsonSchema({
  "type": "object",
  "properties": {
    "string_field": {
      "type": "string"
    },
    "int_field": {
      "type": "integer"
    },
    "array_field": {
      "type": "array",
      "items": {
        "type": "string"
      },
      "additionalItems": false
    },
    "timestamp_field": {
      "type": "string"
    },
    "obj_field": {
      "type": "object",
      "properties": {
        "sub1": {
          "type": "string"
        },
        "sub2": {
          "type": "string"
        }
      },
      "additionalProperties": false
    },
    "struct_field": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "sub3": {
            "type": "string"
          },
          "sub4": {
            "type": "string"
          },
          "sub5": {
            "type": "object",
            "properties": {
              "sub6": {
                "type": "string"
              },
              "sub7": {
                "type": "string"
              }
            },
            "additionalProperties": false
          },
          "sub8": {
            "type": "array",
            "items": {
              "type": "string"
            }
          }
        },
        "additionalProperties": false
      },
      "additionalItems": false
    }
  },
  "additionalProperties": false
}
)

  ; (async () => {
    const writer = await parquetjs.ParquetWriter.openFile(
      parquetSchema,
      examplePath,
    )

    await writer.appendRow({
      string_field: 'string value',
      int_field: 10,
      timestamp_field: new Date(),

      arrayfield: {
        list: [ { element: 'arrayfield val1' }, { element: 'arrayfield val2' } ],
      },

      obj_field: {
        sub1: 'objfield_sub1 val',
        sub2: 'objfield_sub2 val',
      },

      struct_field: {
        list: [
          {
            element: {
              sub8: {
                list: [ { element: 'val1' }, { element: 'val2' } ],
              },
              sub3: 'structfield_string val',
              sub4: 'structfield_string val',
              sub5: {
                sub6: 'structfield_struct_string1 val',
                sub7: 'structfield_struct_string2 val',
              },
            },
          },
        ],
      },
    })

    await writer.close()
  })()
```

File: lib/fields.ts (hunk 4 above)
Comment (wilwade) [reply lib/fields.ts:58]: Ok. Took a bit to get through it all, but pushed up changes and taking your example into a test as well.

Oddly enough [Tad](https://github.com/antonycourtney/tad) opened the "bad" files without error, but what really helped lock it in was https://github.com/manojkarthick/pqrs

File: test/test-files/address.schema.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (wilwade) [inline test/test-files/address.schema.json:1]: This and the other files inside of `test-files` are just snapshots for testing against.

File: lib/fields.ts (hunk 4 above)
Comment (noxify) [reply lib/fields.ts:58]: Didn't know both tools - thanks for sharing

File: lib/fields.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                    element: {
+                        fields,
+                    },
+                }
+            },
+        },
+    }
+}
+
+export function createListField(type: ParquetType, optionalElement = false, optionalListField = true, elementOptions: FieldDefinition = {}): FieldDefinition {
```
[the remark this answers is not in the document]
Comment (noxify) [reply lib/fields.ts:57]: @shannonwells - I agree to the confused part, but switching the positions for `optionalElement` and `optionalListField` would mean that you have to set the optional state everytime.

I think it makes more sense to keep the positions.

```js
// example - current
schema.createListField('UTF8', true)

// example switched positions
schema.createListField('UTF8', false, true)
```

Not sure if it makes sense to switch from positional args to an object ( not sure how to describe - here an example ):

```js
schema.createListField({
  optionalListField: false,
  type: 'UTF8',
  optionalElement: true,
  options: {...}
})
```

**Review by noxify — APPROVED** [review]
LGTM - Shannon added a note about the order for the `createListField` args - Maybe we should wait with the merge until we have a decision.

File: test/fields.test.ts [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        assert.equal(c.repetitionType, 'OPTIONAL');
+        assert.equal(c.encoding, 'PLAIN');
+        assert.equal(c.compression, 'UNCOMPRESSED');
+        assert.equal(c.rLevelMax, 0);
+        assert.equal(c.dLevelMax, 1);
+        assert.equal(!!c.isNested, false);
+        assert.equal(c.fieldCount, undefined);
+    });
+
+    it("Can use simple list field type", function () {
```
[the remark this answers is not in the document]
Comment (wilwade) [reply test/fields.test.ts:151]: Oops. I moved it down lower, but missed the original one I made for it. Thanks! Removed.

File: lib/fields.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (wilwade) [reply lib/fields.ts:57]: I was back and forth on even having this as an option. Optional list field makes sense, but while I can see reasons for the optional element, it was less obvious.

That said @noxify you're idea works without much modification at all. And I like the pattern much better:

We already have `elementOptions` so moved the element optional to there (where it already is in `FieldDefinition`.)  Then just kept with the `optional` pattern (named such for the simpler key/value).

```
export function createListField(type: ParquetType, optional = true, elementOptions: FieldDefinition = { optional: true }): FieldDefinition {
    return {
        type: 'LIST',
        optional,
        fields: {
            list: {
                repeated: true,
                fields: {
                    element: {
                        ...elementOptions,
                        type,
                    },
                },
            },
        },
    }
}```

**Pull request merged**